lib/stellar/transaction_builder.rb

```diff
@@ -0,0 +1,108 @@
+      base_fee: 100, 
+      time_bounds: nil, 
+      memo: nil
+    )
+      raise ArgumentError, "bad :source_account" unless source_account.is_a?(Stellar::KeyPair)
+      raise ArgumentError, "bad :sequence_number" unless sequence_number.is_a?(Integer) && sequence_number > 0
+      raise ArgumentError, "bad :time_bounds" unless time_bounds.is_a?(Stellar::TimeBounds) || time_bounds.nil?
+      raise ArgumentError, "bad :base_fee" unless base_fee.is_a?(Integer) && base_fee >= 100
+
+      @source_account = source_account
+      @sequence_number = sequence_number
+      @base_fee = base_fee
+      @time_bounds = time_bounds
+      @memo = self.make_memo(memo)
+      @operations = Array.new
+    end
+
+    def build
+      if @time_bounds.nil?
+        raise "TransactionBuilder.time_bounds must be set during initialization or by calling set_timeout"
+      elsif !@time_bounds.min_time.is_a?(Integer) or !@time_bounds.max_time.is_a?(Integer)
+        raise "TimeBounds.min_time and max_time must be Integers"
+      elsif @time_bounds.max_time != 0 and @time_bounds.min_time > @time_bounds.max_time
+        raise "Timebounds.max_time must be greater than min_time"
+      elsif @time_bounds.max_time != 0 and @time_bounds.max_time < Time.now.to_i
+        raise "Timebounds.max_time must be in the future"
+      end
+      tx = Stellar::Transaction.new(
+        source_account: @source_account.account_id,
+        fee: @base_fee * @operations.length,
+        seq_num: @sequence_number,
+        time_bounds: @time_bounds,
+        memo: @memo,
+        operations: @operations,
+        ext: Stellar::Transaction::Ext.new(0)
+      )
+      @sequence_number += 1
+      tx
+    end
+
+    def add_operation(operation)
+      raise ArgumentError, "bad operation" unless operation.is_a? Stellar::Operation
+      @operations.push(operation)
+      self
+    end
+
+    def clear_operations()
+      @operations.clear
+      self
+    end
+
+    def set_sequence_number(seq_num)
+      raise ArgumentError, "bad sequence number" unless seq_num.is_a?(Integer) and seq_num > 0
+      @sequence_number = seq_num
+      self
+    end
+
+    def set_timeout(timeout)
+      if !timeout.is_a?(Integer) or timeout < 0
+        raise ArgumentError, "timeout must be a non-negative integer"
+      end
+
+      if @time_bounds.nil?
+        @time_bounds = Stellar::TimeBounds.new(min_time: 0, max_time: nil)
+      end
+
+      if timeout == 0
+        @time_bounds.max_time = timeout
+      else
+        @time_bounds.max_time = Time.now.to_i + timeout
+      end
+
+      self
+    end
+
+    def set_memo(memo)
+      @memo = make_memo(memo)
+      self
+    end
+
+    def make_memo(memo)
+      case memo
+      when Stellar::Memo ;
+        memo
+      when nil ;
+        Memo.new(:memo_none)
+      when Integer ;
+        Memo.new(:memo_id, memo)
+      when String ;
+        Memo.new(:memo_text, memo)
+      when Array ;
+        t, val = *memo
+        Memo.new(:"memo_#{t}", val)
```

Comment:
It's not obvious to me what this is used for. Could you share an example?

Reply:
This function has copied from /lib/stellar/transaction.rb. It doesn't look like there are any other memo types other than the ones already listed so I'll remove it.

Reply:
I think there are two other forms of memos actually. [`MEMO_ID` and `MEMO_HASH`](https://www.stellar.org/developers/guides/concepts/transactions.html). This must be designed to support the rare cases they are used. I'd keep it if it was in the other class.